Add connected takers feed

klochowicz opened the pull request:
- add a new watch channel with information about connected takers from the maker's actor system
- test the watch channel behaviour with a unit test
- add a list of connected takers to maker's UI (for testing/monitoring)

This PR will enable us to write more elaborate assertions (e.g. reconnection of the same taker & resuming its operation) and also enables us to gauge the length & amount of inteactions with the maker

klochowicz committed:
Provide information about connected takers in a dedicated feed

Publish Vec<TakerId> on a dedicated watch channel and emit it on the event
stream.

Add a test that verifies that the maker updates the feed correctly.

klochowicz committed:
Display connected takers in the maker's UI

Simple list of connected takers (as the UI is mostly used for testing, it
doesn't need to look pretty).

thomaseizinger commented on daemon/src/maker_cfd.rs:
```diff
@@ -40,7 +40,11 @@ pub struct NewOrder {
+
+pub struct TakerRemoved {
     pub id: TakerId,
 }
```
Maybe `TakerConnected` and `TakerDisconnected` would be better names?

klochowicz replied:
thanks for the suggestion, they are better. Will change the names & handlers.

thomaseizinger commented on daemon/src/maker_cfd.rs:
```diff
@@ -345,6 +355,21 @@ where
             .await?;
 
+        if !self.connected_takers.insert(taker_id) {
+            anyhow::bail!("Taker already connected: {:?}", &taker_id);
```
Is this really a failure as such?

klochowicz replied:
It's not a failure, but an indication that something is not right with our code....

I've downgraded both to `tracing::warn!`, ideally I'd probably add a `debug_assert!` there to catch such inconsistencies early if they'd happen.

thomaseizinger commented on daemon/src/maker_cfd.rs:
```diff
@@ -345,6 +355,21 @@ where
+
+    async fn handle_taker_removed(&mut self, taker_id: TakerId) -> Result<()> {
+        if !self.connected_takers.remove(&taker_id) {
+            anyhow::bail!("Removed unknown taker: {:?}", &taker_id);
```
Same here, perhaps logging is fine here?
Would make the function idempotent which is good I think.

klochowicz replied:
converted to log.

thomaseizinger commented on daemon/tests/happy_path.rs:
```diff
@@ -133,3 +133,22 @@ async fn taker_notices_lack_of_maker() {
+        Vec::<TakerId>::new(),
+        next(maker.connected_takers_feed()).await.unwrap()
+    );
+}
```
Pretty nice!

klochowicz replied:
thanks :)

klochowicz committed:
Use projection actor to publish information about connected takers

Note: Projection actor is used across both taker and maker, so it really should
    publish information about connected counterparty.
    Taker should use that channel for deriving maker online status.

thomaseizinger commented on daemon/src/lib.rs:
```diff
@@ -1,14 +1,16 @@
 use futures::future::RemoteHandle;
 use maia::secp256k1_zkp::schnorrsig;
+use maker_cfd::TakerDisconnected;
+use model::TakerId;
```
These imports are a bit odd, should be part of the others.

klochowicz replied:
thanks, tidied up the commits - `TakerId` should have been added earlier to ensure the tests run (as well as missing import for `TimeStamp`, as tests would not have compiled.

thomaseizinger commented on maker-frontend/src/MakerApp.tsx:
```diff
@@ -41,6 +42,8 @@ export default function App() {
     const walletInfo = useLatestEvent<WalletInfo>(source, "wallet");
     const priceInfo = useLatestEvent<PriceInfo>(source, "quote");
+    const takersOrUndefined = useLatestEvent<TakerId[]>(source, "takers");
+    let takers = takersOrUndefined ? takersOrUndefined! : [];
```
There is a shortcut for this:

```suggestion
    let takers = takersOrUndefined || [];
```
